plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java

```diff
@@ -177,7 +179,9 @@ private Stream<DeltaLakeSplit> getSplits(
 
         MetadataEntry metadataEntry = tableHandle.getMetadataEntry();
         boolean isOptimize = tableHandle.isOptimize();
-        Set<Map<String, Optional<String>>> partitionsWithAtMostOneFile = isOptimize ? findPartitionsWithAtMostOneFile(validDataFiles) : ImmutableSet.of();
+        // Materialize the valid data files iterator to a list only in case of dealing with OPTIMIZE to avoid rewriting partitions containing only one file
```

Comment:
You could do this lazily as well. Stream add file entries from the checkpoint until you see two files that are small enough to qualify for compaction in the same partition, queue those two files, remember that partition for the future so any other files in it can be queued immediately.

Not required for this PR but a good next step.

Reply:
Good idea.
Refactored the code to queue initially the add file entries until we have at least two files to process.